### PR TITLE
Fix hardcoded limit to 30 in database statement

### PR DESCRIPTION
**Pull request by Toflar**
I have no clue where this hardcoded limit to 30 items is coming from but I don't think this makes any sense. Why would we hardcode a max of 30 if a developer provides a limit `<= 0`? 

Reproducing the issue is easy:

1. Open demo.contao.org
2. Edit multiple on the 3 news within `Contao Demo News`
3. Use the "Copy multiple" operation and copy the news until you have more than 30 news
4. Edit the news list: https://demo.contao.org/contao?do=themes&table=tl_module&act=edit&id=63&rt=cac681e0aff23cc577c9e28af681d.PEqqf3T6nxUDwUk1ywbfzf-yqhXNNg6nWPBae8gR2Ng.VCnGSSO92iFtlz1yjzeQrK7e61b6dXTjGaAdL45_s6BXJekoApHud0qRDA
5. Set `Number of items` to `0`
6. Set `Items per page` to `0`
7. Set `Skip items` to `1`
8. Count the news items on https://demo.contao.org/news.html (`document.querySelectorAll('.mod_newslist div.layout_latest').length`)

Expected output: All news items but the first one.
What's happening instead: The first one is skipped correctly, but a random limit of 30 is applied.

### CODE REVIEW

**Review by fritzmg — DISMISSED** [review]

**Review by bytehead — DISMISSED** [review]

**Review by m-vo — DISMISSED** [review]

**Comment by m-vo** [comment]
In case someone is interested - this was the commit adding the magic number: https://github.com/contao/core/commit/2dcea027a3252c44e1272c469f7d4e43761e7fc9

So I think it's relatively safe to assume that there is no special reason for the value.

**Comment by leofeyer** [comment]
> Why would we hardcode a max of 30 if a developer provides a limit `<= 0`?

Because MySQL does not accept negative values as limit. And `0` is a special case that we never used:

> `LIMIT 0` quickly returns an empty set. This can be useful for checking the validity of a query. It can also be employed to obtain the types of the result columns within applications that use a MySQL API that makes result set metadata available.

See https://dev.mysql.com/doc/refman/8.4/en/limit-optimization.html

So `0` does not mean **all** rows but **no** rows at all, therefore it is wrong to use `PHP_INT_MAX` in this case. Also, we obviously cannot change this in a bugfix release because people will run into memory and execution time limits when we suddenly return `PHP_INT_MAX` rows instead of `30` rows!

**Comment by m-vo** [comment]
> So 0 does not mean all rows but no rows at all

But then the current behavior would also be a bug, because 30 rows are returned. How would someone now run into memory issues if they did non intend to return no results (as this did not work before)?

**Comment by leofeyer** [comment]
The limit is now set to `30` if you pass `0` as argument. If we merge the PR, it will be set to `9_223_372_036_854_775_807` instead. It is pretty obvious to me how this could cause memory issues.

**Comment by Toflar** [comment]
Yes, if as a developer I set `->limit(0, 1)` then I **expect** to run into memory issues if there are 4 billion rows. Because I set the limit to `0`. Why would Contao force `30`?

**Comment by m-vo** [comment]
This is how I see it:
- If you wanted zero results before, there was no way.
- If you wanted all the results before, it was already necessary to use an arbitrary (high) number.
- If you wanted exactly 30 results you could theoretically have specified this by using 0 (instead of 30) but this would be very strange to me.

**Comment by aschempp** [comment]
What happens if you only set an offset in the Doctrine query builder?

**Comment by fritzmg** [comment]
> Also, we obviously cannot change this in a bugfix release because people will run into memory and execution time limits when we suddenly return PHP_INT_MAX rows instead of 30 rows!

This only applies if you have an offset defined. With no offset you already get _all_ rows.

**Comment by fritzmg** [comment]
Just to illustrate the issue in a different way:

```php
// This fetches all news entries
NewsModel::findBy(['1=1'], [], []);

// This fetches 30 news entries
NewsModel::findBy(['1=1'], [], ['offset' => 1]);
```

The reason why we need a limit at all when only defining an offset is because you always need to specify `LIMIT <offset>,<limit>` or `LIMIT <limit> OFFSET <offset>` [in MySQL](https://dev.mysql.com/doc/refman/8.4/en/select.html#idm46097367876304). i.e. you cannot define an offset without a limit. Even MySQL's own documentation says to use a large number if you want to retrieve all records while using an offset:

> To retrieve all rows from a certain offset up to the end of the result set, you can use some large number for the second parameter.

MySQL's example uses `18446744073709551615`, but for simplicity `PHP_INT_MAX` should be enough.

**Comment by fritzmg** [comment]
> What happens if you only set an offset in the Doctrine query builder?

It [also will use](https://github.com/doctrine/dbal/blob/7a8252418689feb860ea8dfeab66d64a56a64df8/src/Platforms/AbstractMySQLPlatform.php#L55) `LIMIT 18446744073709551615 OFFSET 1`, similar to the example in the MySQL documentation.

```php
$this->createQueryBuilder('…')
    ->setFirstResult(1)
    ->getQuery()
    ->getSQL()
;

// SELECT … FROM … LIMIT 18446744073709551615 OFFSET 1
```

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php [hunk 1]
```diff
@@ -197,7 +197,8 @@ public function limit($intRows, $intOffset=0)
 	{
 		if ($intRows <= 0)
 		{
-			$intRows = 30;
+			// Ensure offset works if > 0 and limit is set to 0
+			$intRows = PHP_INT_MAX;
```
Comment (fritzmg) [inline core-bundle/src/Resources/contao/library/Contao/Database/Statement.php]:
```suggestion
			$intRows = '18446744073709551615';
```

Should we use this hard coded value instead after all, just like in [`doctrine/dbal`](https://github.com/doctrine/dbal/blob/7a8252418689feb860ea8dfeab66d64a56a64df8/src/Platforms/AbstractMySQLPlatform.php#L55)?

Comment (Toflar) [reply core-bundle/src/Resources/contao/library/Contao/Database/Statement.php]: Fine with me :)

**Comment by leofeyer** [comment]
> > What happens if you only set an offset in the Doctrine query builder?
> 
> It [also will use](https://github.com/doctrine/dbal/blob/7a8252418689feb860ea8dfeab66d64a56a64df8/src/Platforms/AbstractMySQLPlatform.php#L55) `LIMIT 18446744073709551615 OFFSET 1`, similar to the example in the MySQL documentation.

Okay, then let‘s do the same.

**Comment by fritzmg** [comment]
> Okay, then let‘s do the same.

The PR is already updated with that 👍

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php [hunk 2]
```diff
@@ -197,7 +197,8 @@ public function limit($intRows, $intOffset=0)
 	{
 		if ($intRows <= 0)
 		{
-			$intRows = 30;
+			// Ensure offset works if > 0 and limit is set to 0
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/Database/Statement.php]:
```suggestion
			// 2^64-1 is the maximum of unsigned BIGINT, the biggest limit possible
```

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @Toflar.

**Pull request merged**